pkg/safeguards/safeguards.go

```diff
@@ -0,0 +1,311 @@
+				}
+
+				for _, dFile := range dfDir {
+					df, err := fetchDeploymentFile(path.Join(innerPath, dFile.Name()))
+					if err != nil {
+						return c, dfMap, fmt.Errorf("fetching test deployment: %w", err)
+					}
+					dfMap[removeYAMLExtension(dFile.Name())] = df
+				}
+
+			}
+		}
+
+		constraints = append(constraints, d...)
+	}
+
+	for _, con := range constraints {
+		fullConstraintDir := path.Join(fullPath, con.Name())
+		b, err := os.ReadFile(fullConstraintDir)
+		if err != nil {
+			return c, dfMap, fmt.Errorf("reading constraint file: %s", con.Name())
+		}
+
+		var constraintFile ConstraintFile
+		if err := yaml.Unmarshal(b, &constraintFile); err != nil {
+			return c, dfMap, fmt.Errorf("unmarshaling constraint: %w", err)
+		}
+
+		c = append(c, constraintFile)
+	}
+
+	return c, dfMap, nil
+}
+
+// sanitizeRegoPolicy removes problematic lines from our rego code for consumption as our rego policy in the evalution step
+func sanitizeRegoPolicy(rp string) string {
+	lines := strings.Split(rp, "\n")
+	var newLines []string
+	for _, line := range lines {
+		if !strings.Contains(line, "import data.lib.") && !strings.Contains(line, "package lib.") {
+			newLines = append(newLines, line)
+		}
+	}
+	return strings.Join(newLines, "\n")
+}
+
+// appendLibs appends every lib item from the constraint YAML, separated by newlines
+func appendLibs(libs []string) string {
+	l := ""
+	if len(libs) > 0 {
+		for _, lib := range libs {
+			l += lib + "\n"
+		}
+	}
+
+	return l
+}
+
+func buildQueryString(name string) string {
+	return "x = data." + name + ".violation"
+}
+
+func evaluateQuery(ctx context.Context, file ConstraintFile, dfMap map[string]map[string]interface{}) error {
+	queryString := buildQueryString(file.Metadata.Name)
+
+	// thbarnes: throw in a check for if length is 0 or >1 and error if so
+	l := appendLibs(file.Spec.Targets[0].Libs)
+	regoString := file.Spec.Targets[0].Rego
+	regoPolicy := sanitizeRegoPolicy(regoString + l)
+
+	r := rego.New(
+		rego.Query(queryString),
+		rego.Module("main.rego", regoPolicy))
+
+	query, err := r.PrepareForEval(ctx)
+	if err != nil {
+		return fmt.Errorf("creating rego query: %w", err)
+	}
+
+	// our input state
+	// build inputs PER CONSTRAINT
+
+	input := buildInput(dfMap[file.Name])
+
+	// thbarnes: investigate if you can make your own `input` struct and pass it into here
+	rs, err := query.Eval(ctx, rego.EvalInput(input))
+	if err != nil {
+		return fmt.Errorf("evaluating query: %w", err)
+	}
+
+	fmt.Println("Result:", rs[0].Bindings["x"])
+
+	return nil
+}
+
+// ValidateDeployment is what will be called by `draft validate` to validate the user's deployment manifest
+// against each safeguards constraint
+func ValidateDeployment(df, constraint string) error {
+	// thbarnes: ConstraintsBuilderB will eventually take over
+	ctx := context.TODO()
+
+	var cf ConstraintFetcher
+	cba := ConstraintsBuilderA{}
+	cf = cba
+
+	constraintFiles, dfMap, err := cf.Fetch()
```

Comment:
I think `dfMap` is a map of deployment objects? 2 requests:
1. Can you rename this to be more specific?
2. Can you change the type to a `map[string]appsv1.Deployment`?

Reply:
2 is done

Reply:
refactored so that this is no longer relevent